rak811.go

```diff
@@ -147,180 +167,155 @@ func (l *Lora) Close() {
-	}
-	start := time.Now()
-	reader := bufio.NewReader(l.port)
-	for {
-		resp, err = reader.ReadString('\n')
-		resp = strings.TrimSuffix(strings.TrimSpace(resp), "\r")
-		if err == nil && resp != "" {
-			switch resp {
-			case JoinSuccess:
-				return JoinSuccess, nil
-			case JoinFail:
-				return JoinFail, nil
-			case JoinTimeout:
-				return JoinTimeout, nil
-			default:
-				return "", errors.Errorf("invalid join response resp:%v", resp)
-			}
-		}
-		if time.Since(start) > timeout {
-			return "", errors.New("no response after the set timeout")
-		}
+		return "", fmt.New(resp) // Convert the resp to an error so that the caller handle it properly.
 	}
+	// The module doesn't accept any other command before it returns a response
+	// so need to wait for it.
+	return l.tr(1)
 }
 
-// JoinABP join the configured network in ABP mode
+// JoinABP join the configured network in ABP mode.
 func (l *Lora) JoinABP() (string, error) {
-	return l.tx("join=abp")
+	return l.txr("join=abp", 1)
 }
 
-// Signal check the radio rssi, snr, update by latest received radio packet
+// Signal check the radio rssi, snr, update by latest received radio packet.
 func (l *Lora) Signal() (string, error) {
-	return l.tx("signal")
+	return l.txr("signal", 1)
 }
 
-// GetDataRate get next send data rate
+// GetDataRate get next send data rate.
 func (l *Lora) GetDataRate() (string, error) {
-	return l.tx("dr")
+	return l.txr("dr", 1)
 }
 
-// SetDataRate set next send data rate
+// SetDataRate set next send data rate.
 func (l *Lora) SetDataRate(datarate string) (string, error) {
-	return l.tx(fmt.Sprintf("dr=%s", datarate))
+	return l.txr(fmt.Sprintf("dr=%s", datarate), 1)
 }
 
-// GetLinkCnt get LoRaWAN uplink and downlink counter
+// GetLinkCnt get LoRaWAN uplink and downlink counter.
 func (l *Lora) GetLinkCnt() (string, error) {
-	return l.tx("link_cnt")
+	return l.txr("link_cnt", 1)
 }
 
-// SetLinkCnt set LoRaWAN uplink and downlink counter
+// SetLinkCnt set LoRaWAN uplink and downlink counter.
 func (l *Lora) SetLinkCnt(uplinkCnt, downlinkCnt float32) (string, error) {
-	return l.tx(fmt.Sprintf("link_cnt=%f,%f", uplinkCnt, downlinkCnt))
+	return l.txr(fmt.Sprintf("link_cnt=%f,%f", uplinkCnt, downlinkCnt), 1)
 }
 
 // GetABPInfo
 func (l *Lora) GetABPInfo() (string, error) {
-	return l.tx("abp_info")
+	return l.txr("abp_info", 1)
 }
 
-// Send send data to LoRaWAN network
+// Send send data to LoRaWAN network.
+// The parent function should check for:
+// STATUS_TX_UNCOMFIRMED when sending unconfirmed packets,
+// STATUS_TX_COMFIRMED when sending confirmed packets.
 func (l *Lora) Send(data string) (string, error) {
-	resp, err := l.tx(fmt.Sprintf("send=%s", data))
+	resp, err := l.txr(fmt.Sprintf("send=%s", data), 1)
 	if err != nil {
 		return "", err
 	}
 	if resp != OK {
-		return "", errors.Errorf("invalid send request response resp:%v", resp)
+		return "", fmt.New(resp) // Convert the resp to an error so that the caller handle it properly.
```

Comment:
Use `errors.New(resp)` where errors is from stdlib.

Reply:
hm in that case why not use `errors` for everything and  `errors.Wrapf` where needed like I had in a previous commit?

I thought you wanted to use `fmt` for all errors to keep the imported packages to a minimum?

Reply:
Because `Wrapf` is not the standard lib as `fmt.New` does not exist

Reply:
aah yes , sorry I got confused with the packages and did some global replace that cased this.

I have now also fixed all tests with the latest code as well.